### PR TITLE
Initial version of the TFS Pull Request support Addin

### PR DESCRIPTION
**Pull request by pascalberger**

### CODE REVIEW

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        [CakeMethodAlias]
+        [CakeAliasCategory(CakeAliasConstants.PullRequestSystemCakeAliasCategory)]
+        public static IPullRequestSystem TfsPullRequests(
+            this ICakeContext context,
+            TfsPullRequestSettings settings)
+        {
+            context.NotNull(nameof(context));
+            settings.NotNull(nameof(settings));
+
+            return new TfsPullRequestSystem(context.Log, settings);
```
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs]: How to dispose properly?

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+            else if (!string.IsNullOrWhiteSpace(settings.SourceBranch))
+            {
+                this.Log.Verbose("Read pull request for branch {0}", settings.SourceBranch);
+                var pullRequestSearchCriteria =
+                    new GitPullRequestSearchCriteria()
+                    {
+                        Status = PullRequestStatus.Active,
+                        SourceRefName = settings.SourceBranch
+                    };
+                this.pullRequest =
```
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Add empty line.

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+            var request =
+                this.gitClient.GetThreadsAsync(
+                    this.pullRequest.Repository.Id,
+                    this.pullRequest.PullRequestId,
+                    null,
+                    null,
+                    null,
+                    CancellationToken.None);
+            var threads = request.Result;
```
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Add empty line.

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        private int GetCodeFlowLatestIterationId()
+        {
+            var request =
+                this.gitClient.GetPullRequestIterationsAsync(
+                    this.pullRequest.Repository.Id,
+                    this.pullRequest.PullRequestId,
+                    null,
+                    null,
+                    CancellationToken.None);
+            var iterations = request.Result;
```
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Add empty line.

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+
+            if (iterations == null)
+            {
+                throw new PrcaException("Could not retrieve the iterations");
+            }
+
+            var iterationId = iterations.Max(x => x.Id ?? -1);
+            this.Log.Verbose("Dermined iteration id: {0}", iterationId);
```
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Type: Determined

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+                this.gitClient.GetPullRequestIterationChangesAsync(
+                    this.pullRequest.Repository.Id,
+                    this.pullRequest.PullRequestId,
+                    iterationId,
+                    null,
+                    null,
+                    null,
+                    null,
+                    CancellationToken.None);
+            var changes = request.Result;
```
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Add empty line.

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs (hunk 1 above)
Comment (stffabi) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs]: Either we give this control over to the cake script.
  - Then we need to add `IDisposable` to `IPullRequestSystem`. Having it there may result in the case that some PR-systems don't need it.

Or we could create the connection in TfsPullRequestSystem on demand for the 3 callback functions.
  - By changing the signature of `MarkThreadAsFixed` to receive an enumerable of all threads to get fixed we would only open the connection for 4 times. This shouldn't result in an immense overhead, but on the other hand we could correctly dispose the connections.

File: src/Cake.Prca.PullRequests.Tfs.Tests/packages.config [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
```
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs.Tests/packages.config]: Added #3

File: src/Cake.Prca.PullRequests.Tfs/GitPullRequestCommentThreadExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+            if (thread.Properties == null)
+            {
+                throw new InvalidOperationException("Properties collection is not created.");
+            }
+
+            if (thread.Properties.ContainsKey(CommentSourcePropertyName))
+            {
+                thread.Properties[CommentSourcePropertyName] = value;
+            }
```
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/GitPullRequestCommentThreadExtensions.cs]: Oops, added else block

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs (hunk 2 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Done

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs (hunk 3 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Done

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs (hunk 4 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Done

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs (hunk 5 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Done

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs (hunk 6 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystem.cs]: Done

File: src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs (hunk 1 above)
Comment (pascalberger) [inline src/Cake.Prca.PullRequests.Tfs/TfsPullRequestSystemAliases.cs]: Connection is now created on demand. And created https://github.com/cake-contrib/Cake.Prca/issues/7 for changing signature of `MarkThreadAsFixed`

**Pull request merged**